### PR TITLE
Fix/elu alpha validation

### PR DESCRIPTION
**Pull request by kaivalya-cyber**

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for the contribution! The goal of ensuring `alpha >= 0` to prevent unstable gradients or inverted landscapes is valid. However, there are a few important aspects to consider before this can be merged:

#### 1. Impact on Existing Tests and Backwards Compatibility
The validation strictly blocks `alpha < 0`. However, existing test suites (e.g., in `elu_test.py` or downstream Keras tests) might be explicitly testing negative `alpha` values (e.g., `alpha = -1.0`) to verify mathematical correctness or legacy behavior.
If we add this validation, any such tests will fail. This PR does not update or remove those tests.
Furthermore, this might be a **breaking change** for users who are currently using negative `alpha` values (even if mathematically unusual) or have saved models with negative `alpha`.

#### 2. Serialization and Model Loading
Adding validation to `__init__` means that if a user attempts to load a saved model (e.g., via `from_config` or loading a SavedModel/H5 file) that was configured with a negative `alpha` (before this check was added), the loading process will fail with a `ValueError`.
**Action:** We need to confirm if we want to block loading legacy models with negative `alpha`. If we do, this is a clear breaking change. If we don't, we might need to relax the validation or handle it differently during deserialization.

#### 3. Incomplete Validation for Tensors
The validation logic currently uses `tensor_util.is_tf_type(alpha)` to skip validation if `alpha` is a Tensor:
```python
if not tensor_util.is_tf_type(alpha) and alpha < 0:
```
This means that if a user passes a negative Tensor (e.g., `alpha = tf.constant(-1.0)`), it will bypass this validation entirely.
If the goal is to enforce $\alpha \ge 0$, this validation is incomplete.
**Action:** If Tensors are supported as `alpha`, consider adding dynamic validation (e.g., using `tf.debugging.assert_greater_equal`) or clarify this limitation in the documentation. If Tensors are NOT expected to be passed as configuration parameters here (which is typical for this static validation layer), we might not need the `is_tf_type` guard at all, or we should handle it more rigorously.

#### 4. Missing Unit Tests
The PR introduces new validation logic but does not include corresponding unit tests to verify that it works as expected (e.g., that it raises `ValueError` for negative floats but allows valid ones).
**Action:** Please add unit tests to cover the new validation logic.

Here is an example of tests you might want to add:

```python
import tensorflow as tf

class ELUValidationTest(tf.test.TestCase):
    def test_elu_alpha_validation(self):
        # Valid alpha
        tf.keras.layers.ELU(alpha=0.0)
        tf.keras.layers.ELU(alpha=1.0)
        
        # Invalid alpha
        with self.assertRaisesRegex(ValueError, "should be >= 0"):
            tf.keras.layers.ELU(alpha=-1.0)
        
        # Tensor alpha (bypasses validation in current PR)
        # Note: If we want to fully enforce alpha >= 0, this should also raise an error (dynamically).
        tf.keras.layers.ELU(alpha=tf.constant(-1.0))
```

#### Summary
The PR needs to address potential test failures with negative `alpha`, incomplete validation for Tensors, serialization compatibility, and missing unit tests for the new logic.

**Comment by kaivalya-cyber** [comment]
@dmiltr3 thanks for the thorough review! Addressed all points:
 
1. **KerasTensor guard**: Added `backend.is_keras_tensor(alpha)` to prevent
   TypeError on symbolic KerasTensors during `alpha < 0` evaluation.
 
2. **Runtime assertion for Tensors**: Added `check_ops.assert_non_negative()`
   for `tf.Tensor`/`tf.Variable` alpha — negative Tensor values now raise
   `InvalidArgumentError` at runtime instead of silently passing.
 
3. **Unit tests**: Added `advanced_activations_test.py` covering:
   - Positive scalar (0.0, 1.0, 0.5) — passes
   - Negative scalar (-1.0) — raises ValueError
   - Positive Tensor — passes
   - Negative Tensor — raises InvalidArgumentError at runtime (graph + eager)
   - Positive Variable — passes
   - Zero Tensor boundary — passes
   - None alpha — raises ValueError
 
4. **Backwards compat**: Acknowledged. Raising ValueError for negative alpha
   in __init__ is an intentional breaking change — negative alpha produces
   mathematically invalid ELU behavior. Saved models with negative alpha
   will fail to load, which is correct.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
During internal testing of your changes, we encountered a few issues.

Here is a summary of the issues and suggested improvements:

## 1. `backend.is_keras_tensor` Exception Handling
**Issue:** `backend.is_keras_tensor(alpha)` expects a symbolic tensor instance. If `alpha` is a standard Python scalar (e.g., `float` or `int`), this function raises a `ValueError` instead of returning `False`. This causes crashes when valid scalar values are passed to the layer.

**Fix:** Wrap the check in a try-except block or handle scalar types explicitly before calling `is_keras_tensor`.

## 2. Assertion Execution and Type Safety
**Issue:** `check_ops.assert_non_negative` returns an `Operation` (or triggers statically). If you reassign `alpha = check_ops.assert_non_negative(...)`, `alpha` becomes an `Operation` in graph mode, which fails later in `backend.cast_to_floatx` which expects a tensor-like object. Additionally, in graph mode, assertions must be explicitly attached to the execution graph via `control_dependencies` to ensure they are executed.

**Fix:** Use `ops.control_dependencies` to ensure the assert op is executed, and re-attach the dependency to the `alpha` tensor (e.g., via `array_ops.identity` or a dummy math operation).

## 3. Static vs Runtime Assertions in Tests
**Issue:** In `advanced_activations_test.py`, tests using `constant_op.constant(-1.0)` trigger static assertions during graph construction time (raising `InvalidArgumentError` immediately), not at runtime during `sess.run`.

**Fix:** Update tests to catch the exception during layer instantiation or use placeholders for runtime validation tests.

---

# Suggested Code Changes

## `advanced_activations.py`

```python
    # Securely check if alpha is a tensor without crashing on scalars
    is_tensor = tensor_util.is_tf_type(alpha)
    if not is_tensor:
      try:
        is_tensor = backend.is_keras_tensor(alpha)
      except ValueError:
        is_tensor = False

    if not is_tensor:
      if alpha < 0:
        raise ValueError('The alpha value of an ELU layer '
                         'should be >= 0, got %s' % alpha)
    elif tensor_util.is_tf_type(alpha):
      # Ensure assert is executed in graph mode
      with ops.control_dependencies([
          check_ops.assert_non_negative(
              alpha,
              message='The alpha value of an ELU layer should be >= 0')
      ]):
        # Re-assign alpha to ensure dependency is carried over
        # Note: array_ops.identity is preferred here if available
        alpha = math_ops.add(alpha, 0.0)
```

## `advanced_activations_test.py`

Update `test_elu_alpha_tensor_negative_fails_runtime` to catch the error during layer instantiation if using constants:

```python
  @test_util.run_deprecated_v1
  def test_elu_alpha_tensor_negative_fails(self):
    """ELU layer with negative tf.Tensor alpha should fail."""
    with ops.Graph().as_default():
      with self.assertRaisesRegex(
          errors.InvalidArgumentError, 'should be >= 0'):
        layer = advanced_activations.ELU(
            alpha=constant_op.constant(-1.0))
        with self.session() as sess:
          sess.run(variables.global_variables_initializer())
          sess.run(layer.alpha)
```

And consider adding a test for dynamic tensors (placeholders) to verify runtime assertion:

```python
  @test_util.run_deprecated_v1
  def test_elu_alpha_placeholder_call_graph(self):
    """ELU layer call should work with placeholder alpha in graph mode."""
    with ops.Graph().as_default():
      from google3.third_party.tensorflow.python.ops import array_ops
      alpha_placeholder = array_ops.placeholder(dtypes.float32)
      layer = advanced_activations.ELU(alpha=alpha_placeholder)
      inputs = constant_op.constant([-1.0, 1.0])
      outputs = layer(inputs)
      with self.session() as sess:
        sess.run(variables.global_variables_initializer())
        sess.run(outputs, feed_dict={alpha_placeholder: 2.0})
```
